File: csharp/1383.maximum-performance-of-a-team.cs
/*
 * @lc app=leetcode id=1383 lang=csharp
 *
 * [1383] Maximum Performance of a Team
 */

using System;
using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {
    public int MaxPerformance(int n, int[] speed, int[] efficiency, int k) {
        var engineers = speed
            .Zip(efficiency, (s, e) => (Speed: s, Efficiency: e))
            .Select((t, i) => (Speed: t.Speed, Efficiency: t.Efficiency, Index: i))
            .OrderByDescending(t => t.Efficiency)
            .ToArray();
        var sorted = new SortedSet<(int Speed, int Efficiency, int Index)>(Comparer<(int Speed, int Efficiency, int Index)>.Create((a, b) => {
            if (a.Speed != b.Speed) {
                return a.Speed.CompareTo(b.Speed);
            }
            if (a.Efficiency != b.Efficiency) {
                return a.Efficiency.CompareTo(b.Efficiency);
            }
            return a.Index.CompareTo(b.Index);
        }));
        long ret = 0, speedSum = 0;
        foreach (var engineer in engineers) {
            speedSum += (long)engineer.Speed;
            sorted.Add(engineer);
            if (sorted.Count > k) {
                var min = sorted.Min;
                speedSum -= (long)min.Speed;
                sorted.Remove(min);
            }
            ret = Math.Max(ret, (long)engineer.Efficiency * speedSum);
        }
        return (int)(ret % 1000000007);
    }
}
// @lc code=end

File: java/1383.maximum-performance-of-a-team.java
/*
 * @lc app=leetcode id=1383 lang=java
 *
 * [1383] Maximum Performance of a Team
 */

import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.Queue;

// @lc code=start

class Solution {
    public int maxPerformance(int n, int[] speed, int[] efficiency, int k) {
        Engineer[] engineers = new Engineer[n];
        for (int i = 0; i < n; ++i) {
            engineers[i] = new Engineer(efficiency[i], speed[i]);
        }
        Arrays.sort(engineers, java.util.Comparator.comparingInt(Engineer::efficiency).reversed());
        Queue<Integer> speeds = new PriorityQueue<>(k + 1);
        long ret = 0, sum = 0;
        for (var engineer : engineers) {
            speeds.offer(engineer.speed());
            sum += engineer.speed();
            if (speeds.size() > k) {
                sum -= speeds.poll();
            }
            ret = Math.max(ret, sum * engineer.efficiency());
        }
        return (int)(ret % 1000000007);
    }

    private record Engineer(int efficiency, int speed) {}
}
// @lc code=end

